***
# **<center>COURS PYTHON 2IMACS #1</center>**
# ***<center>Introduction</center>***
# ***<center>Types de Données et Operations</center>***
# ***<center>Structures Conditionnelles</center>***
***

# 1-1 Déroulement de l'enseignement


Au total 8 Séances de  3h45 (3 x 1h15)

- Les 5 premières séances seront des alternances de cours et d'exercices sur machine
- Les 3 dernières seront dédiées à une évaluation de type QCM (25% de la note) et à un projet (75% de la note)


# 1-2 Introduction au langage Python

Python a été créé par Guido van Rossum, un développeur hollandais, et son équipe bénévole. Le développement de Python a commencé en décembre 1989 et la première version publique a été publiée en février 1991.   
Le nom est inspiré par la série télévisée britannique "Monty Python's Flying Circus". Python a été conçu dans le but de créer un langage de programmation qui soit facile à lire, à apprendre et à utiliser.   
Guido van Rossum a travaillé successivement chez Google, Dropbox et Microsoft, mais le développement du langage Python s'est poursuivi en parallèle, en tant que projet bénévole. Python est distribué sous une licence open source.  

Python est un langage de haut niveau, ce qui signifie que sa syntaxe est plus proche du "langage humain" que du "langage machine" comme le serait un langage de bas niveau, comme l'asssembleur par exemple.

### Intérêt d'apprendre Python pour les physiciens et les électroniciens
- Apprentissage relativement simple (en comparaison à d'autres langages)  
- Multiplateforme (le code sera utilisable sous Windows, Linux, Mac...)
- Large communauté (nombreuses bibliothèques et ressources en ligne)
- Dans nos métiers, nous avons de grandes quantités de données experimentales à exploiter, Python est adapté à ce type de tâches.
- C'est le langage de référence pour l'intelligence artificielle
- Open source, permet souvent de s'affranchir de logiciels propriétaires coûteux
- Polyvalent (Les utilisations vont de la science des données, au developpement web, developpement logiciel...)

### Python est un langage interprété (et non compilé)
- Le programme sera "traduit" pour être exécuté par la machine à chaque lancement.
- Avantages : rapidité de développement et portabilité (sans compilation, peu importe l'OS)
- Défauts : plus lent à l'exécution et moins pratique à déployer
- Possibilité d'appeler du code écrit dans d'autres languages (script ou bibliothèque)

### Ressources :
- [Documentation officielle Python](https://docs.python.org/fr/3/)
- Livres à la Bib ou B.U.  
    -  Apprenez à programmer en Python (Le Goff)
    -  Python pour la data science : analyser vos données par la pratique avec NumPy, Pandas, Matplotlib et Seaborn (Velt)
    -  Python pour la data science (Jake VanderPlas)
    -  ...
- Forums ([Stack Overflow](https://stackoverflow.com/) ou autre)
- [GitHub](https://github.com/)
- Outils d'IA Générative

### Version de  Python :

- La version actuelle est python3. Testons la version installée sur nos machines:

In [1]:
!python --version

Python 3.12.1


# 1-3 Types de données et opérations

## 1-3-1 Déclaration de variables

Les variables sont des "conteneurs" qui permettent de stocker des valeurs.
Une variable peut contenir différents types de données, ici des nombres. Python a un typage dynamique, ce qui signifie que l'on a pas besoin de déclarer explicitement son type. Il sera défini à partir de la valeur que l'on va affecter à la variable.

In [2]:
a = 58
b,c = 89,22
# Afficher les valeurs des variables
print(a)
print(b)
print(c)

58
89
22


Nous avons utilisé **print()** pour afficher les valeurs de ces variables, on peut aussi afficher du texte avec des **" "**.   
Attention: si on affiche plusieurs éléments (par exemple du texte et des variables), il faut les séparer avec des virgules.

In [3]:
print('a = ',a)
print('b = ',b)
print('c = ',c)

a =  58
b =  89
c =  22


## 1-3-2 Type de variable entier ou flottant

Déclarer un entier (int: integer):

In [4]:
d = 58

Déclarer un nombre à virgule (float) :

In [5]:
e = 108.89

Vérifier les types:

In [6]:
print('d: ',type(d))
print('e:',type(e))

d:  <class 'int'>
e: <class 'float'>


Faire une opération:

In [7]:
f = d + e
print(f)

166.89


*Remarque: On a additionné 2 types differents un int et un float. En python les variables changent facilement de type, c'est pratique mais ça peut
vite amener à faire des erreurs!*

Parfois, il peut être nécessaire de changer le type d'une variable pour effectuer des opérations spécifiques. On peut rentrer la même valeur avec un autre format.

In [8]:
# Forcer un changement de type:
l = 89
print('déclaration sous la forme l =89, type   :',type(l))
l = 89.0
print('déclaration sous la forme l =89.0, type :',type(l))

déclaration sous la forme l =89, type   : <class 'int'>
déclaration sous la forme l =89.0, type : <class 'float'>


Ou forcer un changement de type, comme ici changer le type en  entier (integer) avec **int(variable)** . De la même manière on peut utiliser **float(variable)** pour une conversion en nombre à virgule.

In [9]:
n = 87.8
print('type avant conversion :',type(n))
print('valeur avant conversion :',n)
o = int(n)
print('type après conversion',type(o))
print('valeur après conversion :',o)


type avant conversion : <class 'float'>
valeur avant conversion : 87.8
type après conversion <class 'int'>
valeur après conversion : 87


Sur les int et float, on a accès aux differentes opérations d'addition, soustraction, multiplication, division avec + - * / 

La division offre 3 possibilité:
- division décimale
- division entière
- division modulo

In [10]:
g = 10
h = 3
i = g / h
print('division g/h = : ',i)

j = g // h
print('division entiere g//h = : ',j)

k = g % h
print('division modulo g//h = : ',k,\
      'ici on trouve le reste de la division',
      '(très utile pour vérifier la parité par exemple)')


division g/h = :  3.3333333333333335
division entiere g//h = :  3
division modulo g//h = :  1 ici on trouve le reste de la division (très utile pour vérifier la parité par exemple)


## 1-3-3 Type de variable chaîne de caractère

Une chaine de caractère est une séquence de caractères qui peuvent être des lettres, chiffres ou symboles.  Pour la déclarer, on utilisera des guillemets.

In [11]:
ma_chaine1 = "Monty" # Guillemets doubles
ma_chaine2 = 'Python' # Ou guillemets simples

In [12]:
print(type(ma_chaine1))
print(type(ma_chaine2))
print(ma_chaine1)
print(ma_chaine2)

<class 'str'>
<class 'str'>
Monty
Python


Les opérations que l'on pourra réaliser sur les chaines seront differentes de celles que l'on peut faire sur des int ou des floats.

In [13]:
valeur_1 = '1'
valeur_2 = '2'
somme = valeur_1 + valeur_2
print('1+2 = ',somme)

1+2 =  12


Le signe + pour des chaines de caractères est une concatenation et non une addition

Mesurer le nombre de caractères dans une chaîne avec **len()** (pour length) qui renverra un interger:

In [14]:
longueur = len(ma_chaine1)
print(type(longueur))
print(longueur)

<class 'int'>
5


Remplacer un élément par un autre avec **replace()**:

In [15]:
ma_chaine3 = 'Monty Python'
print(ma_chaine3)
ma_chaine4 = ma_chaine3.replace('Monty','This is ')
print(ma_chaine4)

Monty Python
This is  Python


En indiquant une valeur, entre crochets **[ ]** à la fin de la chaine on peut lire le caractère se trouvant à l'emplacement (appelé indice) demandé.

In [16]:
print(ma_chaine3[0])
print(ma_chaine3[1])
print(ma_chaine3[2])

M
o
n


*Remarque: Les indices commencent toujours à zéro et non à 1*

On peut accéder aux derniers éléments avec un indice négatif.

In [17]:
print(ma_chaine3[-1])
print(ma_chaine3[-2])
print(ma_chaine3[-3])

n
o
h


Ou a une tranche en donnant les indices de début et de fin séparés par ':' :

In [18]:
print('chaine entière:.......................:',ma_chaine3)
print('entre indice 2 et 8...................:',ma_chaine3[2:8])
print('du début à indice 9:..................',ma_chaine3[:9])  
print('de indice 3 à la fin.................:',ma_chaine3[3:])  
print('entre indice 2 et 8 avec pas de sur 2:',ma_chaine3[2:8:2]) 

chaine entière:.......................: Monty Python
entre indice 2 et 8...................: nty Py
du début à indice 9:.................. Monty Pyt
de indice 3 à la fin.................: ty Python
entre indice 2 et 8 avec pas de sur 2: nyP


On a donc une syntaxe de la forme **chaine[start : stop : step]**:  
start : indice de début de la découpe (inclus)  
stop : indice de fin de la découpe (exclus)  
step : pas, c'est-à-dire le nombre d'indices à sauter à chaque étape (par défaut, step est égal à 1)  

Séparer sur un caractere donné avec **split()**:

In [19]:
# Séparer la chaine sur les caractères espace
print(ma_chaine3.split(' '))

['Monty', 'Python']


Trouver l'emplacement nommé index d'un élément avec **index()**. Les index commencent à zéro:

In [20]:
print(ma_chaine3.index('P'))

6


Quand on veut afficher du texte et des valeurs numériques ensemble, il peut être nécessaire de faire une mise en forme:

In [21]:
personnage = "Sir Robin"
nombre_chevaliers = 100/3

- Mise en forme de texte option 1: Concaténation de chaines avec **+** et séparation par des **,**:

In [22]:
print( personnage + " a vaincu " ,nombre_chevaliers , "chevaliers.")

Sir Robin a vaincu  33.333333333333336 chevaliers.


*Remarque: le resultat n'est pas 33.333333333333333 car  les nombres réels sont représentés sur un nombre fini de bits provoquant des erreurs d'arrondis*

- Mise en forme de texte option 2:

Avec %d, on spécifie le type entier et avec %s le type string  

In [23]:
print("%s a vaincu %f chevaliers." % (personnage, nombre_chevaliers))

Sir Robin a vaincu 33.333333 chevaliers.


**%s** est remplacé par la valeur de la variable personnage.  
**%f** est remplacé par la valeur de la variable nombre_chevaliers.  
Le **%** est utilisé pour formater la chaîne avec les variables spécifiées dans (personnage, nombre_chevaliers).

- Mise en forme de texte option 3: Avec la syntaxe **"f"** et **{ }**

In [24]:
print(f"{personnage} a vaincu {nombre_chevaliers} chevaliers.")

Sir Robin a vaincu 33.333333333333336 chevaliers.


{personnage} est remplacé par la valeur de la variable personnage.  
{nombre_chevaliers} est remplacé par la valeur de la variable nombre_chevaliers.  

Le f avant les guillemets indique une chaîne de format f-string en Python. Cela permet d'insérer des variables directement dans la chaîne en utilisant la syntaxe {}.

Cette dernière syntaxe est la derniere introduite (Python3.6), pour des raisons de performance, de simplicité et de flexibilité, on privilegiera celle ci.

- Pour les float, on peut choisir le nombre de décimales à afficher:

In [25]:
print(f"{personnage} a vaincu {nombre_chevaliers:.2f} chevaliers.")

Sir Robin a vaincu 33.33 chevaliers.


Ici on  choisira probablement l'affichage:

In [26]:
print(f"{personnage} a vaincu {nombre_chevaliers:.0f} chevaliers.")

Sir Robin a vaincu 33 chevaliers.


Récapitulatif des differents spécificateurs de type:

| Spécificateur de type | Description                           |
|-----------------------|---------------------------------------|
| `%s`                  | Chaîne de caractères                   |
| `%d`                  | Nombre entier                |
| `%f`                  | Nombre à virgule flottante             |
| `%e` ou `%E`          | Nombre à virgule flottante en notation scientifique (exponentielle) |
| `%g` ou `%G`          | Nombre à virgule flottante en notation courte (automatique) |
| `%x` ou `%X`          | Nombre entier en base hexadécimale     |         


[Exercices 0 à 4](exercices/Exercices1.ipynb)

## 1-3-4 Entrées / Sorties

La fonction **print()** est utilisée en Python pour afficher des messages ou des valeurs à l'écran. Elle prend en entrée les éléments à afficher.

La fonction **input()** est utilisée pour obtenir une entrée de l'utilisateur. Elle affiche un message facultatif pour guider l'utilisateur et attend que celui-ci saisisse une valeur. La valeur saisie est renvoyée par la fonction et peut être stockée dans une variable.

In [27]:
nom = input("Entrez votre nom : ")
print('Hello',nom)

Entrez votre nom :  simon


Hello simon


# 1-4 Structures Conditionnelles

Les structures conditionnelles permettent d'exécuter des blocs de code en fonction de conditions. Les conditions sont évaluées et le bloc de code correspondant est exécuté si la condition est vraie.
Ce bloc doit être indenté, c'est à dire décalé vers la droite avec 4 espaces.

## 1-4-1 Conditions

In [38]:
a = 8
b = 8
c = 5

- Condition sur l'égalité avec un double égal (**==**) : 

In [39]:
if a == b:
    print('a égal b')

a égal b


On peut prévoir une alternative au cas où la condition ne serait pas vérifiée en utilisant le mot clé **'else':**

In [40]:
if a == c:
    print('a = c')
else:
    print('a different de c')

a different de c


Ou plusieurs alternatives au cas où la condition précédente ne serait pas vérifiée, en utilisant le mots clé **'elif'** avant un **'else'** final. **elif** est une contraction de **else if**

In [41]:
if a == 1:
    print('a = 1')
elif a == 2:
    print('a = 2')
elif a == 3:
    print('a = 3')
else:
    print('a > 3')

a > 3


- Condition sur la difference avec **!=** ou avec **not**:

In [42]:
if a != c:
    print('a different de c')
# autre notation:
if not a == c:
    print('a égal c')

a different de c
a égal c


La syntaxe **!=** sera généralement préférée, la syntaxe **not** risque de rendre le code moins lisible et sera généralement évitée.

- Double condition:

Les deux conditions doivent être respectées (**and**)

In [43]:
if a == 8 and c == 5:
    print('a = 8 et c = 5')

a = 8 et c = 5


Au moins une des conditions doit être respectée (**or**)

In [44]:
if a == 16 or c == 5:
    print('a = 8 ou c = 5')

a = 8 ou c = 5


- Condition sur un opérateur booleen

In [45]:
condition1 = (a==b)
print(condition1)
if condition1:
    print('a = b')

True
a = b


*Remarque: On a ici un nouveau type de variable avec True, c'est un booleen*

In [46]:
print(type(condition1))

<class 'bool'>


Récapitulatif des opérateurs:

| Opérateur | Définition                                         |
| --------- | --------------------------------------------------- |
| ==        | Permet de tester l'égalité en valeur et en type     |
| !=        | Permet de tester la différence en valeur ou en type |
| <         | Permet de tester si une valeur est strictement inférieure à une autre |
| >         | Permet de tester si une valeur est strictement supérieure à une autre |
| <=        | Permet de tester si une valeur est inférieure ou égale à une autre |
| >=        | Permet de tester si une valeur est supérieure ou égale à une autre |

Et des booleens:

| Opérateur | Définition                                    |
|-----------|-----------------------------------------------|
| `and`     | Renvoie `True` si les deux expressions sont vraies |
| `or`      | Renvoie `True` si au moins une des expressions est vraie |
| `not`     | Inverse le résultat de l'expression logique     |



# 1-5 obligations et bonnes pratiques syntaxiques 

Autant que possible respecter les règles de [PEP8](https://peps.python.org/pep-0008/)

Quelques points essentiels sont résumés ici:
- Utilisez des noms de variables et de fonctions significatifs, en suivant la convention snake_case (en minuscules avec des underscores).
- Indentez votre code avec des espaces, en utilisant 4 espaces par niveau d'indentation.
- Limitez la longueur de vos lignes de code à 79 caractères maximum.
- Utilisez des espaces autour des opérateurs (=, +, -, etc.) et après les virgules dans les listes et les appels de fonction.
- Évitez d'utiliser des espaces inutiles à la fin des lignes ou à la fin des expressions.
- Utilisez des commentaires pour expliquer votre code lorsque cela est nécessaire, mais assurez-vous qu'ils sont pertinents et à jour.
- Importez les modules sur des lignes séparées et placez-les au début du script.
- Évitez d'utiliser des noms de variables qui ressemblent à des mots-clés Python existants.
- Utilisez des espaces après les virgules dans les tuples, les listes et les dictionnaires.
- Séparez les définitions de fonctions et les classes avec deux lignes vides.
- Favorisez la réutilisation de code en évitant les répétitions et en créant des fonctions lorsque cela est approprié. Suivez le principe DRY (Don't Repeat Yourself) cela améliore la lisibilité, la maintenabilité et la flexibilité de votre code, tout en facilitant les modifications et les mises à jour ultérieures.


Tim Peters, un des principaux developpeur du langage a même intégré au langage une commande emettant un certain nombre de recommandations de manière assez poètique...

In [47]:
import this

[Exercices Structures Conditionnelles: 5 à 11](exercices/Exercices1.ipynb) 